components/suggestion/command_provider/app_command_parser/app_command_parser.ts

```diff
@@ -0,0 +1,1256 @@
+        const call = parsed.form?.call || parsed.binding?.call || parsed.binding?.form?.call;
+        const hasRequired = this.getMissingFields(parsed).length === 0;
+        const hasValue = (parsed.state !== ParseState.EndValue || (parsed.field && parsed.values[parsed.field.name] !== undefined));
+
+        if (executableStates.includes(parsed.state) && call && hasRequired && hasValue) {
+            const execute = getExecuteSuggestion(parsed);
+            if (execute) {
+                suggestions = [execute, ...suggestions];
+            }
+        }
+
+        return suggestions.map((suggestion) => this.decorateSuggestionComplete(parsed, suggestion));
+    }
+
+    // composeCallFromParsed creates the form submission call
+    composeCallFromParsed = async (parsed: ParsedCommand): Promise<AppCallRequest | null> => {
+        if (!parsed.binding) {
+            return null;
+        }
+
+        const call = parsed.form?.call || parsed.binding.call;
+        if (!call) {
+            return null;
+        }
+
+        const values: AppCallValues = parsed.values;
+        const ok = await this.expandOptions(parsed, values);
+
+        if (!ok) {
+            return null;
+        }
+
+        const context = this.getAppContext(parsed.binding.app_id);
+        return createCallRequest(call, context, {}, values, parsed.command);
+    }
+
+    expandOptions = async (parsed: ParsedCommand, values: AppCallValues) => {
+        if (!parsed.form?.fields) {
+            return true;
+        }
+
+        let ok = true;
+        await Promise.all(parsed.form.fields.map(async (f) => {
+            if (!values[f.name]) {
+                return;
+            }
+            switch (f.type) {
+            case AppFieldTypes.DYNAMIC_SELECT:
+                values[f.name] = {label: '', value: values[f.name]};
+                break;
+            case AppFieldTypes.STATIC_SELECT: {
+                const option = f.options?.find((o) => (o.value === values[f.name]));
+                if (!option) {
+                    ok = false;
+                    this.displayError(this.intl.formatMessage({
+                        id: 'apps.error.command.unknown_option',
+                        defaultMessage: 'Unknown option for field `{fieldName}`: `{option}`.',
+                    }, {
+                        fieldName: f.name,
+                        option: values[f.name],
+                    }));
+                    return;
+                }
+                values[f.name] = option;
+                break;
+            }
+            case AppFieldTypes.USER: {
+                let userName = values[f.name] as string;
+                if (userName[0] === '@') {
+                    userName = userName.substr(1);
+                }
+                let user = selectUserByUsername(this.store.getState(), userName);
+                if (!user) {
+                    const dispatchResult = await this.store.dispatch(getUserByUsername(userName) as any);
```

Comment:
(0/5) Obviously not something to change now since it would be really involved, but I'm curious why the client is doing this validation. Since we're still needing to call out to the server when doing the validation for some of these fields, client-side validation seems less helpful here

Reply:
@mickmister can you please comment?

Reply:
I believe this is to make sure the `CallRequest` is formed properly before sending to the server. This may be something applicable to be handled in the proxy. @larkox Do you have thoughts on this?

Reply:
Yes, this is extra validation on client side. The best thing about validation on client side is that we are 100% sure it is scoped to the permissions of the client (cannot reach outside the team, etc...). Also, this simplifies the live of the App developer, since they are receiving always a proper user. Finally, it removes weight from the proxy so it can just relay the message instead of doing extra validation.

The general case here is that the user is already in the redux store (because among other things, it was taken from the select). If it is not, it might be because the user copy pasted the command and the user is not yet in redux. That is why we get the User from the server. If still there is no user there, is because the user have introduced badly the user (copy paste or not selecting from the suggestions). All that information is kind of "already on the client side" so there is no need to make a trip to the server to do the validation.

If we leave the validation to the proxy, we would need either:
- Make the proxy aware of the binding and the form so they can validate/fill the missing information (mainly user id)
- Mark the fields that need validation somehow, specifying whether they are users or channels.